security/cmd/rdist/doc/rdist.man
.\"
.\" Copyright (c) 1983 Regents of the University of California.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	$Id: rdist.man,v 1.1 1996/01/31 23:28:11 dkarlton Exp $
.\"	@(#)rdist.1	6.6 (Berkeley) 5/13/86
.\"
.TH RDIST 1 "March 14, 1994"
.UC 6
.SH NAME
rdist \- remote file distribution client program
.SH SYNOPSIS
.B rdist
[ 
.B \-DFn
]
[ 
.B \-A 
.I num
] 
[ 
.B \-a 
.I num 
]
[ 
.B \-d 
.I var=value 
] 
[ 
.B \-l
.I <local logopts>
] 
[ 
.B \-L
.I <remote logopts>
] 
[ 
.B \-f 
.I distfile 
] 
[ 
.B \-M 
.I maxproc
] 
[ 
.B \-m 
.I host 
]
[ 
.B \-o
.I distopts
]
[ 
.B \-t 
.I timeout 
] 
[
.B \-p
.I <rdistd-path>
]
[
.B \-P
.I <rsh-path>
]
[ 
.I name ...
]
.PP
.B rdist
.B \-DFn
.B -c 
.I name ... 
.I [login@]host[:dest]
.PP
.B rdist
.B \-Server
.PP
.B rdist
.B \-V
.SH DESCRIPTION
.I Rdist
is a program to maintain identical copies of files over multiple hosts. 
It preserves the owner, group, mode, and mtime of files if possible and
can update programs that are executing.
.I Rdist
reads commands from
.I distfile
to direct the updating of files and/or directories.
If
.I distfile
is `\-', the standard input is used.
If no
.B \-f
option is present, the program looks first for `distfile',
then `Distfile' to use as the input.
If no names are specified on the command line,
.I rdist
will update all of the files and directories listed in
.IR distfile .
Otherwise, the argument is taken to be the name of a file to be updated
or the label of a command to execute. If label and file names conflict,
it is assumed to be a label.
These may be used together to update specific files
using specific commands.
.PP
The
.B \-c
option forces
.I rdist
to interpret the remaining arguments as a small
.IR distfile .
The equivalent distfile is as follows.
.nf

.ti +.5i
( \fIname\fP ... ) -> [\fIlogin\fP@]\fIhost\fP
.ti +1i
install	[\fIdest\fP] ;

.fi
.PP
The
.B \-Server
option is recognized to provide partial backward compatible support
for older versions of
.I rdist
which used this option to put
.I rdist
into server mode.
If
.I rdist
is started with the 
.B \-Server
command line option, it will attempt to exec (run) the old version of
.I rdist.
This option will only work if 
.I rdist
was compiled with the location of the old rdist
(usually either
.I /usr/ucb/oldrdist
or
.I /usr/old/rdist)
and that program is available at run time.
.PP
.I Rdist
can use either the
.I rcmd(3)
function call or the 
.I rsh(1c),
remote shell, command to access each target host.
The method used is selected at compile-time.
If the
.I rsh(1c)
method is used and
the target host is the string
.B localhost
and 
the remote user name is the same as the local user name,
.I rdist
will run the command
.nf
.sp
.RS
.B "/bin/sh -c rdistd -S"
.RE
.sp
.fi
Otherwise
.I rdist
run will run the command
.nf
.sp
.RS
\fBrsh \fIhost\fB -l \fIremuser \fBrdistd -S\fR
.RE
.sp
.fi
where 
.I host
is the name of the target host,
.I remuser
is the name of the user to make the connection as and,
.I rdistd
is the rdist server command on the target host as shown below.
.PP
If the
.I rcmd(3) 
method is used, then
.I rdist
makes the connection to the target host itself and runs
the 
.I rdistd
server program as shown below.
The default, and preferred method, is to use
.I rsh(1c)
to make the connection to target hosts.  This allows 
.I rdist
to be run without being setuid to ``root''.
.PP
On each target host
.I Rdist
will attempt to run the command
.nf
.sp
.RS
.I "rdistd -S"
.RE
.sp
.fi
or
.nf
.sp
.RS
.I "<rdistd path> -S"
.RE
.sp
.fi
if the
.I \-p
option was specified.
If no
.B \-p 
option is included,
or the
.I <rdistd path>
is a simple filename,
.I rdistd
or
.I <rdistd path>
must be somewhere in the 
.B $PATH
of the user running
.B rdist
on the remote (target) host.
.SH OPTIONS
.TP
.B "\-A \fInum\fR"
Set the minimum number of free files (inodes) on a filesystem that must exist
for 
.I rdist
to update or install a file.
.TP
.B "\-a \fInum\fR"
Set the minimum amount of free space (in bytes) on a filesystem that must exist
for 
.I rdist
to update or install a file.
.TP
.B \-D
Enable copious debugging messages.
.TP
.B "\-d \fIvar=value\fR"
Define
.I var
to have
.IR value .
This
option is used to define or override variable definitions in the
.IR distfile .
.I Value
can be the empty string, one name, or a list of names surrounded by
parentheses and separated by tabs and/or spaces.
.TP
.B \-F
Do not fork any child 
.I rdist
processes.
All clients are updated sequentially.
.TP
.B "\-f \fIdistfile\fR"
Set the name of the distfile to use to be
.I distfile .
If 
.I distfile
is specified as
``\-'' (dash)
then read from standard input (stdin).
.TP
.B "\-l \fIlogopts\fR"
Set local logging options.
See the section 
.B "MESSAGE LOGGING"
for details on the syntax for
.I logopts.
.TP
.B "\-L \fIlogopts\fR"
Set remote logging options.
.I logopts
is the same as for local logging
except the values are passed to the remote
server (\fIrdistd\fR).
See the section 
.B "MESSAGE LOGGING"
for details on the syntax for
.I logopts.
.TP
.B "\-M \fInum\fR"
Set the maximum number of simultaneously
running child
.I rdist
processes to
.I num.
The default is 4.
.TP
.B "\-m \fImachine\fR"
Limit which machines are to be updated. Multiple
.B \-m
arguments can be given to limit updates to a subset of the hosts listed in the
.IR distfile .
.TP
.B \-n
Print the commands without executing them. This option is
useful for debugging
.IR distfile .
.TP
.B "\-o\fIdistopts\fR"
Specify the dist options to enable.
.I distopts
is a comma separated list of options which are listed below.
The valid values for
.I distopts
are:
.RS
.IP \fBverify\fR
Verify that the files are up to date on all the hosts. Any files
that are out of date will be displayed but no files will be changed
nor any mail sent.
.IP \fBwhole\fR
Whole mode. The whole file name is appended to the destination directory
name.  Normally, only the last component of a name is used when renaming files.
This will preserve the directory structure of the files being
copied instead of flattening the directory structure. For example,
rdisting a list of files such as
.I /path/dir1/f1 
and
.I /path/dir2/f2 
to 
.I /tmp/dir 
would create
files 
.I /tmp/dir/path/dir1/f1 
and 
.I /tmp/dir/path/dir2/f2 
instead of 
.I /tmp/dir/dir1/f1 
and 
.I /tmp/dir/dir2/f2.
.IP \fBnoexec\fR
Automatically exclude executable files that are in 
.I a.out(5)
format from being checked or updated. 
.IP \fByounger\fR
Younger mode. Files are normally updated if their
.I mtime
and
.I size
(see
.IR stat (2))
disagree. This
option causes
.I rdist
not to update files that are younger than the master copy.
This can be used
to prevent newer copies on other hosts from being replaced.
A warning message is printed for files which are newer than the master copy.
.IP \fBcompare\fR
Binary comparison. Perform a binary comparison and update files if they differ
rather than comparing dates and sizes.
.IP \fBfollow\fR
Follow symbolic links. Copy the file that the link points to rather than the
link itself.
.IP \fBignlnks\fR
Ignore unresolved links.
.I Rdist
will normally try to maintain the link structure of files being transferred
and warn the user if all the links cannot be found.
.IP \fBchknfs\fR
Do not check or update files on target host that
reside on NFS filesystems.
.IP \fBchkreadonly\fR
Enable check on target host
to see if a file resides on a read-only filesystem.
If a file does, then no checking or updating of the file is attempted.
.IP \fBchksym\fR
If the target on the remote host is a symbolic link, but is not on the
master host, the remote target will be left a symbolic link.
This behavior is generally considered a bug in the original version of
.I rdist,
but is present to allow compatibility with older versions.
.IP \fBquiet\fR
Quiet mode. Files that are being modified are normally
printed on standard output. This
option suppresses this.
.IP \fBremove\fR
Remove extraneous files. If a directory is being updated, any files that exist
on the remote host that do not exist in the master directory are removed.
This is useful for maintaining truly identical copies of directories.
.IP \fBnochkowner\fR
Do not check user ownership of files that already exist.
The file ownership is only set when the file is updated.
.IP \fBnochkgroup\fR
Do not check group ownership of files that already exist.
The file ownership is only set when the file is updated.
.IP \fBnochkmode\fR
Do not check file and directory permission modes.
The permission mode is only set when the file is updated.
.IP \fBnodescend\fR
Do not descend into a directory.
Normally 
.I rdist
will recursively check directories.
If this option is enabled, then any files listed in the
file list in the distfile that are directories are not recursively scanned.
Only the existence, ownership, and mode of the directory are checked.
.IP \fBnumchkgroup\fR
Use the numeric group id (gid) to check group ownership instead of
the group name.
.IP \fBnumchkowner\fR
Use the numeric user id (uid) to check user ownership instead of
the user name.
.IP \fBsavetargets\fR
Save files that are updated instead of removing them.
Any target file that is updates is first rename from
.B file
to
.B file.OLD.
.RE
.TP
.B "\-p \fI<rdistd-path>\fR"
Set the path where the rdistd server is searched for on the target host.
.TP
.B "\-P \fI<rsh-path>\fR"
Set the path to the 
.I rsh(1c)
command.
The
.I rsh-path
may be a colon seperated list of possible pathnames.
In this case, the first component of the path to exist is used.
i.e. 
.B "/usr/ucb/rsh:/usr/bin/remsh",
.B /usr/bsd/rsh.
.TP
.B "\-t \fItimeout\fR"
Set the timeout period (in seconds) for waiting for responses from the remote 
.I rdist
server.
The default is 900 seconds.
.TP
.B \-V
Print version information and exit.
.SH "MESSAGE LOGGING"
.I Rdist
uses a collection of predefined message
.B facilities
that each contain a list of message
.B types
specifying which types of messages to send to that 
.I facility.
The local client (\fIrdist\fR) and the remote server (\fIrdistd\fR) each
maintain
their own copy of what types of messages to log to what facilities.
.LP
The 
.B \-l
.I logopts
option to
.I rdist
tells
.I rdist
what logging options to use locally.
The 
.B \-L
.I logopts
option to
.I rdist
tells
.I rdist
what logging options to pass to the remote
.I rdistd
server.
.LP
The form of
.I logopts
should be of form
.sp
.RS
\fIfacility\fB=\fItypes\fB:\fIfacility\fB=\fItypes...
.RE
.sp
The valid facility names are:
.RS
.IP \fBstdout\fR
Messages to standard output.
.IP \fBfile\fR
Log to a file.  To specify the file name, use the format 
``\fBfile=\fIfilename\fB=\fItypes\fR''.
e.g.
.B "``file=/tmp/rdist.log=all,debug''.
.IP \fBsyslog\fR
Use the 
.I syslogd(8)
facility.
.IP \fBnotify\fR
Use the internal
.I rdist
.B notify
facility.
This facility is used in conjunction with the
.B notify
keyword in a 
.I distfile
to specify what messages are mailed to the
.B notify
address.
.RE
.LP
.I types
should be a comma separated list of message types.  Each message type 
specified enables that message level.  This is unlike the
.I syslog(3)
system facility which uses an ascending order scheme.
The following
are the valid 
.I types:
.RS
.IP \fBchange\fR
Things that change.
This includes files that are installed or updated in some way.
.IP \fBinfo\fR
General information.
.IP \fBnotice\fR
General info about things that change.
This includes things like making directories which are needed in order
to install a specific target, but which are not explicitly specified in
the
.I distfile.
.IP \fBnerror\fR
Normal errors that are not fatal.
.IP \fBferror\fR
Fatal errors.
.IP \fBwarning\fR
Warnings about errors which are not as serious as
.B nerror
type messages.
.IP \fBdebug\fR
Debugging information.
.IP \fBall\fR
All but debug messages.
.RE
.LP
Here is a sample command line option:
.nf
.sp
.RS
\-l stdout=all:syslog=change,notice:file=/tmp/rdist.log=all
.RE
.sp
.fi
This entry will set local message logging to have all but debug
messages sent to standard output, change and notice messages will
be sent to 
.I syslog(3),
and all messages will be written to the file
.B /tmp/rdist.log.
.SH DISTFILES
.PP
The
.I distfile
contains a sequence of entries that specify the files
to be copied, the destination hosts, and what operations to perform
to do the updating. Each entry has one of the following formats.
.nf

.RS
<variable name> `=' <name list>
[ label: ] <source list> `\->' <destination list> <command list>
[ label: ] <source list> `::' <time_stamp file> <command list>
.RE

.fi
The first format is used for defining variables.
The second format is used for distributing files to other hosts.
The third format is used for making lists of files that have been changed
since some given date.
The \fIsource list\fP specifies a
list of files and/or directories on the local host which are to be used
as the master copy for distribution.
The \fIdestination list\fP is the list of hosts to which these files are to be
copied.  Each file in the source list is added to a list of changes
if the file is out of date on the host which is being updated (second format) or
the file is newer than the time stamp file (third format).
.PP
Labels are optional. They are used to identify a command for partial updates.
.PP
Newlines, tabs, and blanks are only used as separators and are
otherwise ignored. Comments begin with `#' and end with a newline.
.PP
Variables to be expanded begin with `$' followed by one character or
a name enclosed in curly braces (see the examples at the end).
.PP
The source and destination lists have the following format:
.nf

.ti +.5i
<name>
or
.ti +.5i
`(' <zero or more names separated by white-space> `)'

.fi
These simple lists can be modified by using one level of set addition,
subtraction, or intersection like this:
.nf

.ti +.5i
list '-' list
or
.ti +.5i
list '+' list
or
.ti +.5i
list '&' list

.fi
If additional modifications are needed (e.g., ``all servers and client
machines except for the OSF/1 machines'') then the list will have
to be explicitly constructed in steps using "temporary" variables.
.PP
The shell meta-characters `[', `]', `{', `}', `*', and `?'
are recognized and expanded (on the local host only) in the same way as
.IR csh (1).
They can be escaped with a backslash.
The `~' character is also expanded in the same way as
.IR csh
but is expanded separately on the local and destination hosts.
When the
.B \-o\fIwhole\fR
option is used with a file name that begins with `~', everything except the
home directory is appended to the destination name.
File names which do not begin with `/' or `~' use the destination user's
home directory as the root directory for the rest of the file name.
.PP
The command list consists of zero or more commands of the following
format.
.nf

.RS
.ta \w'cmdspecial    'u +\w'name list    'u
`install'	<options>	opt_dest_name `;'
`notify'	<name list>	`;'
`except'	<name list>	`;'
`except_pat'	<pattern list>	`;'
`special'	<name list>	string `;'
`cmdspecial'	<name list>	string `;'
.RE

.fi
.PP
The
.I install
command is used to copy out of date files and/or directories.
Each source file is copied to each host in the destination list.
Directories are recursively copied in the same way.
.I Opt_dest_name
is an optional parameter to rename files.
If no
.I install
command appears in the command list or
the destination name is not specified,
the source file name is used.
Directories in the path name will be created if they
do not exist on the remote host.
The
\fB\-o \fIdistopts\fR
option
as specified above under
.B OPTIONS,
has the same semantics as
on the command line except they only apply to the files
in the source list.
The login name used on the destination host is the same as the local host
unless the destination name is of the format ``login@host".
.PP
The
.I notify
command is used to mail the list of files updated (and any errors
that may have occurred) to the listed names.
If no `@' appears in the name, the destination host is appended to
the name
(e.g., name1@host, name2@host, ...).
.PP
The
.I except
command is used to update all of the files in the source list
.B except
for the files listed in \fIname list\fP.
This is usually used to copy everything in a directory except certain files.
.PP
The
.I except_pat
command is like the
.I except
command except that \fIpattern list\fP is a list of regular expressions
(see
.IR ed (1)
for details).
If one of the patterns matches some string within a file name, that file will
be ignored.
Note that since `\e' is a quote character, it must be doubled to become
part of the regular expression.  Variables are expanded in \fIpattern list\fP
but not shell file pattern matching characters.  To include a `$', it
must be escaped with `\e'.
.PP
The
.I special
command is used to specify
.IR sh (1)
commands that are to be executed on the
remote host after the file in \fIname list\fP is updated or installed.
If the \fIname list\fP is omitted then the shell commands will be executed
for every file updated or installed.  
.I String
starts and ends with `"' and can cross multiple lines in
.I distfile.
Multiple commands to the shell should be separated by `;'.

security/cmd/rdist/src/rdist.c
/*
 * Copyright (c) 1983 Regents of the University of California.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef lint
static char RCSid[] = 
"$Id: rdist.c,v 1.2 1996/04/24 00:12:42 dkarlton Exp $";

static char sccsid[] = "@(#)main.c	5.1 (Berkeley) 6/6/85";

static char copyright[] =
"@(#) Copyright (c) 1983 Regents of the University of California.\n\
 All rights reserved.\n";
#endif /* not lint */


#include "defs.h"
#include <netdb.h>
#include <sys/ioctl.h>
#ifdef USE_SSL
#include <termios.h>
#include <stdlib.h>
#endif /* USE_SSL */
/*
 * Remote distribution program.
 */

char   	       *distfile = NULL;		/* Name of distfile to use */
int     	maxchildren = MAXCHILDREN;	/* Max no of concurrent PIDs */
int		nflag = 0;			/* Say without doing */
long		min_freespace = 0;		/* Min filesys free space */
long		min_freefiles = 0;		/* Min filesys free # files */
FILE   	       *fin = NULL;			/* Input file pointer */
struct group   *gr = NULL;			/* Static area for getgrent */
char		localmsglist[] = "stdout=all:notify=all:syslog=nerror,ferror";
char   	       *remotemsglist = NULL;
#ifdef USE_SSL
char		optchars[] = "A:a:bcd:DFf:hiK:k:l:L:M:m:NnOo:p:P:qRrS:st:Vvwxy";
#else /* USE_SSL */
char		optchars[] = "A:a:bcd:DFf:hil:L:M:m:NnOo:p:P:qRrst:Vvwxy";
#endif /* USE_SSL */
FILE   	       *opendist();
char	       *path_rdistd = _PATH_RDISTD;
char	       *path_remsh = _PATH_REMSH;

#ifdef USE_SSL
char	       *secargs = 0; 	/* default to not use ssl rsh */
int		keyflag = 0;	/* defaults to no client auth */
char	       *keydir = 0;	/* no key directory */
char 	       *password = 0;
char	       *nickname = 0;   /* name for key and cert, for client auth */
int             dongleflag = 0;

static void echoOff(int fd)
{
    if (isatty(fd)) {
        struct termios tio;
        tcgetattr(fd, &tio);
        tio.c_lflag &= ~ECHO;
        tcsetattr(fd, TCSAFLUSH, &tio);
    }
}

static void echoOn(int fd)
{
    if (isatty(fd)) {
        struct termios tio;
        tcgetattr(fd, &tio);
        tio.c_lflag |= ECHO;
        tcsetattr(fd, TCSAFLUSH, &tio);
    }
}

static char *getPassword(char *prompt)
{
    FILE *in, *out;
    char buf[2000];

    in = fopen("/dev/tty", "r");
    out = fopen("/dev/tty", "w");
    echoOff(fileno(out));
    for (;;) {
        fprintf(out, prompt);
        fflush(out);
        buf[0] = 0;
        fgets(buf, sizeof(buf), in);
        fprintf(out, "\n");
        if (buf[0]) break;
    }
    echoOn(fileno(out));
    fclose(out);
    fclose(in);
    return strdup(buf);
}

#endif /* USE_SSL */

/*
 * Add a hostname to the host list
 */
static void addhostlist(name, hostlist)
	char *name;
	struct namelist **hostlist;
{
	register struct namelist *ptr, *new;

	if (!name || !hostlist)
		return;

	new = (struct namelist *) xmalloc(sizeof(struct namelist));
	new->n_name = strdup(name);
	new->n_next = NULL;

	if (*hostlist) {
		for (ptr = *hostlist; ptr && ptr->n_next; ptr = ptr->n_next)
			;
		ptr->n_next = new;
	} else
		*hostlist = new;
}

main(argc, argv, envp)
	int argc;
	char *argv[];
	char **envp;
{
	struct namelist *hostlist = NULL;
	register int x;
	register char *cp;
	int cmdargs = 0;
	int c;
	int path_set = 0;
	
	/*
	 * We initialize progname here instead of init() because
	 * things in msgparseopts() need progname set.
	 */
	setprogname(argv);

	if (cp = msgparseopts(localmsglist, TRUE)) {
		error("Bad builtin log option (%s): %s.", 
		      localmsglist, cp);
		usage();
	}

	if (init(argc, argv, envp) < 0)
		exit(1);

	/*
	 * Be backwards compatible.
	 */
	for (x = 1; x <= argc && argv[x]; x++) {
		if (strcmp(argv[x], "-Server") != 0)
			continue;
#if	defined(_PATH_OLDRDIST)
		message(MT_SYSLOG, 
			"Old rdist (-Server) requested; running %s", 
			_PATH_OLDRDIST);
		(void) execl(_PATH_OLDRDIST, basename(_PATH_OLDRDIST), 
			     "-Server", (char *)NULL);
		fatalerr("Exec old rdist failed: %s: %s.", 
			 _PATH_OLDRDIST, SYSERR);
#else	/* !_PATH_OLDRDIST */
		fatalerr("Old rdist not available.");
#endif	/* _PATH_OLDRDIST */
		exit(1);
	}

#if	defined(DIRECT_RCMD)
	if (becomeuser() != 0)
		exit(1);
#else	/* !DIRECT_RCMD */
	/*
	 * Perform check to make sure we are not incorrectly installed
	 * setuid to root or anybody else.
	 */
	if (getuid() != geteuid())
		fatalerr("This version of rdist should not be installed setuid.");
#endif	/* DIRECT_RCMD */

	while ((c = getopt(argc, argv, optchars)) != -1)
		switch (c) {
#ifdef USE_SSL
		        break;
		case 'k':
			keyflag = 1;
			nickname = strdup(optarg);
			break;
		case 'K':
			keydir = strdup(optarg);
			break;
		case 'S':
			secargs = strdup(optarg);
			if ( !path_set ) {
			    /* replace the path if it has not already been
			     * specified on the command line.
			     */
			    path_remsh = _PATH_SREMSH;
			}
			break;
		    
#endif /* USE_SSL */
		case 'l':
			if (cp = msgparseopts(optarg, TRUE)) {
				error("Bad log option \"%s\": %s.", optarg,cp);
				usage();
			}
			break;

		case 'L':
			remotemsglist = strdup(optarg);
			break;

		case 'A':
		case 'a':
		case 'M':
		case 't':
			if (!isdigit(*optarg)) {
				error("\"%s\" is not a number.", optarg);
				usage();
			}
			if (c == 'a')
				min_freespace = atoi(optarg);
			else if (c == 'A')
				min_freefiles = atoi(optarg);
			else if (c == 'M')
				maxchildren = atoi(optarg);
			else if (c == 't')
				rtimeout = atoi(optarg);
			break;

		case 'F':
			do_fork = FALSE;
			break;

		case 'f':
			distfile = strdup(optarg);
			if (distfile[0] == '-' && distfile[1] == CNULL)
				fin = stdin;
			break;

		case 'm':
			addhostlist(optarg, &hostlist);
			break;

		case 'd':
			define(optarg);
			break;

		case 'D':
			debug = DM_ALL;
			if (cp = msgparseopts("stdout=all,debug", TRUE)) {
				error("Enable debug messages failed: %s.", cp);
				usage();
			}
			break;

		case 'c':
			cmdargs++;
			break;

		case 'n':
			nflag++;
			break;

		case 'V':
			printf("%s\n", getversion());
			exit(0);

		case 'o':
			if (parsedistopts(optarg, &options, TRUE)) {
				error("Bad dist option string \"%s\".", 
				      optarg);
				usage();
			}
			break;

		case 'p':
			if (!optarg) {
				error("No path specified to \"-p\".");
				usage();
			}
			path_rdistd = strdup(optarg);
			break;

		case 'P':
			if (!optarg) {
				error("No path specified to \"-P\".");
				usage();
			}
			if (cp = searchpath(optarg)) {
				path_remsh = strdup(cp);
				path_set = 1;
			} else {
				error("No component of path \"%s\" exists.",
				      optarg);
				usage();
			}
			break;

			/*
			 * These options are obsoleted by -o.  They are
			 * provided only for backwards compatibility
			 */
		case 'v':	FLAG_ON(options, DO_VERIFY);		break;
		case 'N':	FLAG_ON(options, DO_CHKNFS);		break;
		case 'O':	FLAG_ON(options, DO_CHKREADONLY);	break;
		case 'q':	FLAG_ON(options, DO_QUIET);		break;
		case 'b':	FLAG_ON(options, DO_COMPARE);		break;
		case 'r':	FLAG_ON(options, DO_NODESCEND);		break;
		case 'R':	FLAG_ON(options, DO_REMOVE);		break;
		case 's':	FLAG_ON(options, DO_SAVETARGETS);	break;
		case 'w':	FLAG_ON(options, DO_WHOLE);		break;
		case 'y':	FLAG_ON(options, DO_YOUNGER);		break;
		case 'h':	FLAG_ON(options, DO_FOLLOW);		break;
		case 'i':	FLAG_ON(options, DO_IGNLNKS);		break;
		case 'x':	FLAG_ON(options, DO_NOEXEC);		break;

		case '?':
		default:
			usage();
		}

#ifdef USE_SSL
	if (keyflag) {
	    char donglefile[MAXPATHLEN];

	    if (!keydir) {
		keydir = getenv("SSL_DIR");
		if (!keydir) {
		    sprintf(donglefile, "/usr/etc/ssl/dongle");
		}
            } else {
	        sprintf(donglefile, "%s/dongle", keydir);
            }

	    if (access(donglefile, R_OK) == 0) {
		dongleflag = 1;
	    }

	    if (!dongleflag) {
		password = getPassword("Key Password: ");
	    }
	}
#endif /* USE_SSL */
	
	if (debug) {
		printf("%s\n", getversion());
		msgprconfig();
	}

	if (nflag && IS_ON(options, DO_VERIFY))
		fatalerr(
		 "The -n flag and \"verify\" mode may not both be used.");

	/*
	 * Don't fork children for nflag
	 */
	if (nflag)
		do_fork = 0;

	if (cmdargs)
		docmdargs(realargc - optind, &realargv[optind]);
	else {
		if (fin == NULL)
			fin = opendist(distfile);
		(void) yyparse();
		/*
		 * Need to keep stdin open for child processing later
		 */
		if (fin != stdin)
			(void) fclose(fin);
		if (nerrs == 0)
			docmds(hostlist, realargc-optind, &realargv[optind]);
	}

	exit(nerrs != 0);
}

/*
 * Open a distfile
 */
FILE *opendist(distfile)
	char *distfile;
{
	char *file = NULL;
	FILE *fp;

	if (distfile == NULL) {
		if (access("distfile", R_OK) == 0)
			file = "distfile";
		else if (access("Distfile", R_OK) == 0)
			file = "Distfile";
	} else {
		/*
		 * Try to test to see if file is readable before running m4.
		 */
		if (access(distfile, R_OK) != 0)
			fatalerr("%s: Cannot access file: %s.", 
				 distfile, SYSERR);
		file = distfile;
	}

	if (file == NULL)
		fatalerr("No distfile found.");

	fp = fopen(file, "r");

	if (fp == NULL)
		fatalerr("%s: open failed: %s.", file, SYSERR);

	return(fp);
}

/*
 * Print usage message and exit.
 */
usage()
{
	char *sopts = "cDFnv";

	(void) fprintf(stderr,
		      "Usage: %s [-%s] [-A <num>] [-a <num>] [-d var=value]\n",
		       progname, sopts);
	(void) fprintf(stderr, 
       "\t[-f distfile] [-l <msgopt>] [-L <msgopt>] [-M <maxproc>]\n");
	(void) fprintf(stderr, 
       "\t[-m host] [-o <distopts>] [-p <rdistd-cmd>] [-P <rsh-path>]\n");
	(void) fprintf(stderr, 
       "\t[-t <timeout>] [target ...]\n");

	(void) fprintf(stderr,
		      "OR:    %s [-%s] -c source [...] machine[:dest]\n", 
		       progname, sopts);

	(void) fprintf(stderr, "OR:    %s -V\n", progname);

	(void) fprintf(stderr, "\nThe values for <distopts> are:\n\t%s\n",
		       getdistoptlist());

	msgprusage();

	exit(1);
}

/*
 * rcp like interface for distributing files.
 */
docmdargs(nargs, args)
	int nargs;
	char *args[];
{
	register struct namelist *nl, *prev;
	register char *cp;
	struct namelist *files, *hosts;
	struct subcmd *cmds;
	char *dest;
	static struct namelist tnl = { NULL, NULL };
	int i;

	if (nargs < 2)
		usage();

	prev = NULL;
	files = NULL;
	for (i = 0; i < nargs - 1; i++) {
		nl = makenl(args[i]);
		if (prev == NULL)
			files = prev = nl;
		else {
			prev->n_next = nl;
			prev = nl;
		}
	}

	cp = args[i];
	if ((dest = strchr(cp, ':')) != NULL)
		*dest++ = '\0';
	tnl.n_name = cp;
	hosts = expand(&tnl, E_ALL);
	if (nerrs)
		exit(1);

	if (dest == NULL || *dest == '\0')
		cmds = NULL;
	else {
		cmds = makesubcmd(INSTALL);
		cmds->sc_options = options;
		cmds->sc_name = dest;
	}

	debugmsg(DM_MISC, "docmdargs()\nfiles = %s", getnlstr(files));
	debugmsg(DM_MISC, "host = %s", getnlstr(hosts));

	insert((char *)NULL, files, hosts, cmds);
	docmds(0, (char **)NULL, 0, (char **)NULL);
}

/*
 * Get a list of NAME blocks (mostly for debugging).
 */
extern char *getnlstr(nl)
	register struct namelist *nl;
{
	static char buf[16384];
	register int count = 0, len = 0;

	(void) sprintf(buf, "(");

	while (nl != NULL) {
		if (nl->n_name == NULL)
			continue;
		len += strlen(nl->n_name) + 2;
		if (len >= sizeof(buf)) {
			(void) strcpy(buf,
				      "getnlstr() Buffer not large enough");
			return(buf);
		}
		++count;
		(void) strcat(buf, " ");
		(void) strcat(buf, nl->n_name);
		nl = nl->n_next;
	}

	(void) strcat(buf, " )");

	return(buf);
}
